src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -222,43 +227,143 @@ def _push_content_single_part(self) -> str:
+            idx_participant = indidividual_rows[0].index('participant_id')  # first row are the fieldnames
+            participant_values = []
+            for row in indidividual_rows:
+                participant_values.append(row[idx_participant])
+            fieldname = 'entity:participant_id'
+            participant =  [fieldname] + list(set(participant_values[1:]))
+
+            # Write participant and sample data to their respective files.
+            tsv_filenames = ['participant.tsv', 'sample.tsv']
+            for tsv_filename in tsv_filenames:
+                with open(os.path.join(tsv_file_dir, tsv_filename), 'w') as f:
+                    if tsv_filename == 'participant.tsv':
+                        f.write(''.join([str(x) + '\n' for x in participant]))
+                    else:
+                        f.write(sample_values)
+                        #copy(os.path.join(tsv_file_dir, tsv_filename), data_path)
+                copy(os.path.join(tsv_file_dir, tsv_filename), data_path)
+
+            assert tsv_filenames == os.listdir(data_path)
+            bag = bdbag_api.make_bag(bag_path, update=True)  # update TSV checksums
+            assert bdbag_api.is_bag(bag_path)
+            bdbag_api.validate_bag(bag_path)
+            assert bdbag_api.check_payload_consistency(bag)
+
+            return bdbag_api.archive_bag(bag_path, 'zip')
+
+    def _iterate_hit_tsv(self, es_search_hit, writer):
         hit_dict = es_search_hit.to_dict()
         assert len(hit_dict['contents']['files']) == 1
         file = hit_dict['contents']['files'][0]
         file_fields = self._translate(file, 'contents.files')
+
         for bundle in hit_dict['bundles']:
             # FIXME: If a file is in multiple bundles, the manifest will list it twice. `hca dss download_manifest`
             # would download the file twice (https://github.com/DataBiosphere/azul/issues/423).
             writer.writerow(self._translate(bundle, 'bundles') + file_fields)
 
+    def _iterate_hit_bdbag(self, es_search_hit, sample_writer):
+        hit_dict = es_search_hit.to_dict()
+
+        assert len(hit_dict['contents']['specimens']) == 1
+        assert len(hit_dict['contents']['cell_suspensions']) == 1
+        assert len(hit_dict['contents']['files']) == 1
+
+        specimen = one(hit_dict['contents']['specimens'])
+        cell_suspension = one(hit_dict['contents']['cell_suspensions'])
+        file = one(hit_dict['contents']['files'])
+
+        specimen_fields = self._translate(specimen, 'contents.specimens')
+        cell_suspension_fields = self._translate(cell_suspension, 'contents.cell_suspensions')
+        file_fields = self._translate(file, 'contents.files')
+
+        # Construct URL for file.
+        file_id = file['uuid']
+        version = file['version']
+        replica = 'gcp'
+        endpoint = f'v1/files/{file_id}?version={version}&replica={replica}'
+        fetch_url = [urljoin(base=config.dss_endpoint, url=endpoint)]
```

Comment:
It's inconsistent to use urljoin but not use https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode for embedding version and replica. We know that version and replica are urlsafe so I'm ok with the latter. I don't think urljoin() actually buys us anything over just concatenating so

```suggestion
        fetch_url = config.dss_endpoint + '/' + endpoint
```

In fact, you mention `v1` in `endpoint` above which duplicates what's already at the end of `config.dss_endpoint` so your use of urljoin() makes this overly complicated.

Reply:
Replaced `urljoin` by simple str concatenation.